inappmessaging/src/test/java/com/rakuten/tech/mobile/inappmessaging/runtime/data/repositories/CampaignRepositorySpec.kt

```diff
@@ -0,0 +1,233 @@
+    }
+
+    @Test
+    fun `should modify impressionsLeft if maxImpressions was updated (campaign modified)`() {
+        CampaignRepository.instance().syncWith(listOf(message0), 0)
+        CampaignRepository.instance().decrementImpressions(message0.getCampaignId())
+        CampaignRepository.instance().decrementImpressions(message0.getCampaignId())
+        CampaignRepository.instance().messages.first().impressionsLeft shouldBeEqualTo 1
+
+        val updatedCampaign = ValidTestMessage(
+            campaignId = message0.getCampaignId(),
+            maxImpressions = 6
+        )
+        CampaignRepository.instance().syncWith(listOf(updatedCampaign), 0)
+        CampaignRepository.instance().messages.first().impressionsLeft shouldBeEqualTo 4
+    }
+
+    // endregion
+
+    // region optOutCampaign
+
+    @Test
+    fun `should mark campaign as opted out`() {
+        CampaignRepository.instance().syncWith(listOf(message0), 0)
+        CampaignRepository.instance().messages.first().isOptedOut?.shouldBeFalse()
+
+        val updatedCampaign = CampaignRepository.instance().optOutCampaign(message0)
+        updatedCampaign?.isOptedOut?.shouldBeTrue()
+        CampaignRepository.instance().messages.first().isOptedOut?.shouldBeTrue()
+    }
+
+    @Test
+    fun `should not cache updated campaign's opt out status if it is marked as isTest`() {
+        val testCampaign = ValidTestMessage(isTest = true)
+        CampaignRepository.instance().syncWith(listOf(testCampaign), 0)
+
+        CampaignRepository.instance().optOutCampaign(testCampaign)
+    }
+
+    @Test
+    fun `should return null when opting out unknown campaign`() {
+        val testCampaign = ValidTestMessage("0")
+        val testCampaign1 = ValidTestMessage("1")
+        CampaignRepository.instance().syncWith(listOf(testCampaign), 0)
+
+        CampaignRepository.instance().optOutCampaign(testCampaign1).shouldBeNull()
+    }
+
+    // endregion
+
+    // region decrementImpressionsLeft
+
+    @Test
+    fun `should decrement impressionsLeft value`() {
+        CampaignRepository.instance().syncWith(listOf(message0), 0)
+        val impressions = message0.impressionsLeft ?: 0
+
+        val updatedCampaign = CampaignRepository.instance().decrementImpressions(message0.getCampaignId())
+        updatedCampaign?.impressionsLeft shouldBeEqualTo impressions - 1
+        CampaignRepository.instance().messages.first().impressionsLeft shouldBeEqualTo impressions - 1
+    }
+
+    @Test
+    fun `should not decrement if impressionsLeft value is already 0`() {
+        val testCampaign = ValidTestMessage(maxImpressions = 0)
+        CampaignRepository.instance().syncWith(listOf(testCampaign), 0)
+        CampaignRepository.instance().decrementImpressions(testCampaign.getCampaignId())
+
+        CampaignRepository.instance().messages.first().impressionsLeft shouldBeEqualTo 0
+    }
+
+    // endregion
+
+    @Test
+    fun `should not update max impression`() {
```

Comment:
A bit confusing. Title is "should not update max.." but the checking in the last line is checking that value is 6 (updated value).

Reply:
Thanks for noticing, updated test name